### ISSUE TITLE
clippy::needless_match automatically applies wrong migration

### ISSUE THREAD

**Issue by mineichen** [rust-lang/rust-clippy#15032]
### Summary

Cargo clippy --fix causes my code to behave differently...
In the example, `println` shouldn't  happen if the input was Error::Baz

### Lint Name

needless_match 

### Reproducer

I tried this code:

```rust
enum Error {
    Bar,
    Baz,
}

fn foo(input: Result<u64, Error>) -> Result<u64, Error> {
    let apply_res = match input {
        x @ (Ok(_) | Err(Error::Bar)) => x,
        e => return e,
    };

    println!("Do something");

    apply_res
}

```
I saw this happen:

```rust

fn foo(input: Result<u64, Error>) -> Result<u64, Error> {
    let apply_res = input;

    println!("Do something");

    apply_res
}


```

I expected to see this happen:
input = Err(Error::Baz) shouldn't cause a println

### Version

```text
rustc 1.87.0 (17067e9ac 2025-05-09)
binary: rustc
commit-hash: 17067e9ac6d7ecb70e50f92c1944e545188d2359
commit-date: 2025-05-09
host: x86_64-pc-windows-msvc
release: 1.87.0
LLVM version: 20.1.1
```

### Additional Labels

* `l-suggestion-causes-error`

**Comment by profetia** [comment]
I cannot reproduce this in the latest nightly. Have you try a newer version of clippy?

**Comment by donkomura** [comment]
I also cannot reproduce this in nightly.
It seems to be resolved in #14757.

**Comment by mineichen** [comment]
You are right. Nightly doesn't cause this error anymore. Sorry for the inconveniences
